ciao-controller/tenants.go

```diff
@@ -69,3 +72,155 @@ func (c *controller) PatchTenant(tenantID string, patch []byte) error {
 	// we need to update through datastore.
 	return c.ds.JSONPatchTenant(tenantID, patch)
 }
+
+func (c *controller) CreateTenant(tenantID string, config types.TenantConfig) (types.TenantSummary, error) {
+	// tenant ID must be a UUID4
+	tuuid, err := uuid.Parse(tenantID)
+	if err != nil {
+		return types.TenantSummary{}, err
+	}
+
+	// SubnetBits must be between 4 and 30
+	if config.SubnetBits == 0 {
+		config.SubnetBits = 24
+	} else {
+		if config.SubnetBits < 4 || config.SubnetBits > 30 {
+			return types.TenantSummary{}, errors.New("subnet bits must be between 4 and 30")
+		}
+	}
+
+	tenant, err := c.ds.AddTenant(tuuid.String(), config)
+	if err != nil {
+		return types.TenantSummary{}, err
+	}
+
+	tenant.CNCIctrl, err = newCNCIManager(c, tenantID)
+	if err != nil {
+		return types.TenantSummary{}, err
+	}
+
+	ts := types.TenantSummary{
+		ID:   tenant.ID,
+		Name: tenant.Name,
+	}
+
+	ref := fmt.Sprintf("%s/tenants/%s", c.apiURL, tenant.ID)
+	link := types.Link{
+		Rel:  "self",
+		Href: ref,
+	}
+	ts.Links = append(ts.Links, link)
+
+	return ts, nil
+}
+
+func (c *controller) deleteInstances(tenantID string) error {
+	// remove any external IPs
+	ips := c.ListMappedAddresses(&tenantID)
+	for _, addr := range ips {
+		err := c.UnMapAddress(addr.ExternalIP)
+		if err != nil {
+			return errors.Wrap(err, "Unable to remove tenant")
+		}
+	}
+
+	// delete all this tenant's instances.
+	instances, err := c.ds.GetAllInstancesFromTenant(tenantID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to remove tenant")
+	}
+
+	for _, i := range instances {
+		err := c.deleteInstanceSync(i.ID)
```

Comment:
You might want to do this in parallel.  Otherwise in a worst case scenario, a node hosting 100 instances from the tenant you're deleting drops off the network, it will take 3 hours and 20 minutes, to delete all the instances.